### PR TITLE
Relearning after unit updating

### LINKED ISSUE #643: GET /query: cancel learning if it's already started
`GET /query` for anomaly detector starts learning if `alpha` / `confidence` changed
Thus, if you change it twice - you get `Can't start learning when it's already started` error
We should cancel learning before starting a new one

### PR DESCRIPTION
**Pull request by amper43**
fixes #643 

The problem is that `AnomalyAnalyticUnit` doesn't need learning after update but it did learn

changes:
- relearn unit after updating only for analytic units where it's needed

### CODE REVIEW

File: server/src/models/analytic_units/utils.ts [hunk 1]
```diff
@@ -13,15 +13,18 @@ export function createAnalyticUnitFromObject(obj: any): AnalyticUnit {
...
     case DetectorType.THRESHOLD:
-      return ThresholdAnalyticUnit.fromObject(obj);
+      return ThresholdAnalyticUnit;
 
     default:
-      throw new Error(`Can't create analytic unit with type "${detectorType}"`);
-  }
-}
+      throw new Error(`Unsupported detector type "${detectorType}"`);
+}
```
Comment (rozetko) [inline server/src/models/analytic_units/utils.ts]:
```suggestion
}

```

File: server/src/models/analytic_units/analytic_unit_model.ts [hunk 2]
```diff
@@ -4,6 +4,9 @@ import { Metric } from 'grafana-datasource-kit';
 
 
 export abstract class AnalyticUnit {
+
+  public needLearnAfterUpdate = false;
```
Comment (rozetko) [inline server/src/models/analytic_units/analytic_unit_model.ts]: `needLearningAfterUpdate` maybe?

Comment (jonyrock) [reply server/src/models/analytic_units/analytic_unit_model.ts]: `learningAfterUpdateRequired`?

File: server/src/models/analytic_units/utils.ts [hunk 3]
```diff
@@ -13,15 +13,18 @@ export function createAnalyticUnitFromObject(obj: any): AnalyticUnit {
   }
 
   const detectorType: DetectorType = obj.detectorType;
+  return getClassByDetectorType(detectorType).fromObject(obj);
+}
+
+export function getClassByDetectorType(detectorType: DetectorType): AnalyticUnit {
```
Comment (jonyrock) [inline server/src/models/analytic_units/utils.ts]: you return here a class, not an instance (see definition). It's better return instance.

File: server/src/models/analytic_units/analytic_unit_model.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ import { Metric } from 'grafana-datasource-kit';
 
 export abstract class AnalyticUnit {
 
-  public needLearnAfterUpdate = false;
+  public learningAfterUpdateRequired= false;
```
Comment (jonyrock) [inline server/src/models/analytic_units/analytic_unit_model.ts]: `public learningAfterUpdateRequired = false;` (space before `=`)

File: server/src/models/analytic_units/anomaly_analytic_unit_model.ts [hunk 5]
```diff
@@ -6,7 +6,7 @@ import { Metric } from 'grafana-datasource-kit';
 
 export class AnomalyAnalyticUnit extends AnalyticUnit {
 
-  public needLearnAfterUpdate = true;
+  public learningAfterUpdateRequired= true;
```
Comment (jonyrock) [inline server/src/models/analytic_units/anomaly_analytic_unit_model.ts]: public learningAfterUpdateRequired = false; (space before =)

File: server/src/models/analytic_units/threshold_analytic_unit_model.ts [hunk 6]
```diff
@@ -15,7 +15,7 @@ export enum Condition {
 
 export class ThresholdAnalyticUnit extends AnalyticUnit {
 
-  public needLearnAfterUpdate = true;
+  public learningAfterUpdateRequired= true;
```
Comment (jonyrock) [inline server/src/models/analytic_units/threshold_analytic_unit_model.ts]: public learningAfterUpdateRequired = false; (space before =)

**Pull request merged**